### ISSUE TITLE
Enforce branch column order and colors

### ISSUE THREAD

**Issue by haizz** [nicoespeon/gitgraph.js#243]
First of all, thank you for the great project!

Currently the order of branch columns is determined by which branch comes first in commit list:

```typescript
export class GraphColumns<TNode> {
  private branches: Set<Branch["name"]> = new Set();

  public constructor(commits: Array<Commit<TNode>>) {
    commits.forEach((commit) => this.branches.add(commit.branchToDisplay));
  }
// ...
}
```

My app has an UI which hides/shows branches based on user requests by filtering commits which get imported to gitgraph, and arbitrarily changing branch order and colors become confusing for users. I'd like to have master branch always having column index 0, first selected branch having column index 1, second branch having column index 2, ... regardless of which commit comes first in the log. 

What would be the easiest way to accomplish this?

**Comment by haizz** [comment]
BTW, I'm using `import` function from `GitgraphUserApi`.

**Comment by fabien0102** [comment]
You can specify the color in "manual" mode with `gitgraph.branch({name: "mybranch", style: {color: "pink"}})` -> https://codesandbox.io/embed/ym2jrq408j (and the dot color is not following ^^ time to open a new issue 😅 )

For the `import`, you don't have (yet) any option to customize this, but it's really interesting problem ^^
I guess we can add a `branchToColor` options somewhere for this kind of case.

```tsx
<Gitgraph options={{branchToColor: (branch) => "black"}}></Gitgraph>
```

**Comment by haizz** [comment]
Will that enforce the column order? What if I would like to keep recent branches closer to the master? Currently it's strictly earlier commits who receive lower column index.

**Comment by fabien0102** [comment]
mmmm, you're right, maybe a `sortBranch` 🤔 with the same API as the native `Array.sort`

**Comment by haizz** [comment]
I can make a PR if you define an API and throw me some basic guidelines on where/how to implement this.

**Comment by haizz** [comment]
Maybe just pass an array of branch names with desired order?

**Comment by fabien0102** [comment]
Sure but in the case of the `import` how do you know what branch name you have? same in the classic mode, the config is the first thing you give, so seams a bit fragile to give an array of branch names.

**Comment by fabien0102** [comment]
I don't exactly remember how where is defined the branch order, I need to dig a bit in the code 😅 

As guideline, all the project is tested (unit tests with jest for the `core`, visual tested with storybook for `react` and `js`), so depending where is the problem, we always begin to add a test or a story to demonstrate the bug, and after fix it! (TDD style ^^)

For this kind of feature, everything should be on `gitgraph-core`, and `getRenderedData` (https://github.com/nicoespeon/gitgraph.js/blob/master/packages/gitgraph-core/src/__tests__/gitgraph.getRenderedData.branchesPaths.test.ts)

Definitely not the most trivial part but it well tested 😉

**Comment by haizz** [comment]
```typescript
type BranchOrderFunction = (branchNameA: string, branchNameB: string) => number;

interface GitgraphOptions {
    //...
    branchOrderFunction?: BranchOrderFunction;
}
```

or

```typescript
type BranchOrderFunction<TNode> = (branchA: Branch<TNode>, branchB: Branch<TNode>) => number;
```

?

**Comment by nicoespeon** [comment]
Branch color depends on "branches paths" ordering. Branches paths are computed from commit positions. And commits position depends on the computed "column":

https://github.com/nicoespeon/gitgraph.js/blob/aa07aa76fce60214aa4ecdfd3468950d470ca058/packages/gitgraph-core/src/gitgraph.ts#L194-L199

This is what you're looking for. 

In short: the column of the commit determines the order of the rendered branch.

The one responsible of determining the column is https://github.com/nicoespeon/gitgraph.js/blob/master/packages/gitgraph-core/src/graph-columns.ts

For the moment, branches are inserted in the order they appear in commits history. We could change this logic and apply the sorting function you are talking about.

As I'm not sure what would be the different use-cases, I'd start with providing less information to the end-user (exposing less is better to keep flexibility). Thus, I'd go for the first option: **provide branch names only**:

```ts
type BranchOrderFunction = (branchNameA: string, branchNameB: string) => number;
```

---

Regarding the data structure used in `GraphColumns` to store branches, it's currently a `Set`. It might not be convenient for sorting. If you need to, you can change it.

---

I did a quick experiment changing the sorting order to get `develop` before `master` and here's the result:

![image](https://user-images.githubusercontent.com/1094774/55048129-780e7d80-501d-11e9-9f81-67b8b0c7ff4b.png)

Though, I needed to fix branch color calculation:

https://github.com/nicoespeon/gitgraph.js/blob/aa07aa76fce60214aa4ecdfd3468950d470ca058/packages/gitgraph-core/src/gitgraph.ts#L232-L239

It doesn't use computed column currently—but, it's another issue than the one you noticed @fabien0102. 

I'll clean that fix and push it later.

**Comment by nicoespeon** [comment]
I just pushed a change, so it's not anymore `GraphColumns` but `BranchesOrder`. I hope it expresses the intention better.

**Comment by haizz** [comment]
@nicoespeon @fabien0102 thank you for your help! I've made a PR.

**Comment by nicoespeon** [comment]
Thanks for your contribution @haizz 👋 

As mentioned in the PR: I simply refactor the name to be more explicit. I went for the semantics of `Array.prototype.sort()`, thus I called it `CompareBranchesOrder` instead of `BranchesOrderFunction`.

I published **v1.1.0-2** of @gitgraph/react & @gitgraph/core that contains the change. Let's know if you find some issues with different ordering of branches.
Thanks!
